openmls/tests/book_code.rs

```diff
@@ -1663,3 +1663,176 @@ fn commit_builder() {
+        bob_provider,
+        &bob_signature_keys,
+    );
+
+    // === Alice adds Bob ===
+    let (_mls_message_out, welcome, _group_info) = alice_group
+        .add_members(
+            alice_provider,
+            &alice_signature_keys,
+            &[bob_key_package.key_package().clone()],
+        )
+        .expect("Could not add members.");
+
+    let welcome: MlsMessageIn = welcome.into();
+
+    let group_context: Option<GroupContext> =
+        bob_provider.storage().group_context(&group_id).unwrap();
+    assert!(group_context.is_none());
+    let confirmation_tag: Option<ConfirmationTag> =
+        bob_provider.storage().confirmation_tag(&group_id).unwrap();
+    assert!(confirmation_tag.is_none());
+
+    let state_before = GroupStorageState::from_storage(bob_provider.storage(), &group_id);
+    //ANCHOR: not_join_group_welcome
+    let welcome = match welcome.extract() {
+        MlsMessageBodyIn::Welcome(welcome) => welcome,
+        _ => unimplemented!("Handle other message types"),
+    };
+    //ANCHOR_END: not_join_group_welcome
+
+    //ANCHOR: not_join_group_welcome_inspect
+    for _secret in welcome.secrets() {
+        // TODO: check secret
+    }
+    //ANCHOR_END: not_join_group_welcome_inspect
+
+    //ANCHOR: not_join_group_processed_welcome
+    let join_config = MlsGroupJoinConfig::default();
+    let processed_welcome = ProcessedWelcome::new_from_welcome(bob_provider, &join_config, welcome)
+        .expect("Error constructing processed welcome");
+    //ANCHOR_END: not_join_group_processed_welcome
+
+    //ANCHOR: not_join_group_processed_welcome_inspect
+    let _unverified_psks = processed_welcome.psks();
+    let _unverified_group_info = processed_welcome.unverified_group_info();
+    //ANCHOR_END: not_join_group_processed_welcome_inspect
+
+    //ANCHOR: not_join_group_staged_welcome
+    let staged_welcome: StagedWelcome = processed_welcome
+        .into_staged_welcome(bob_provider, None)
+        .expect("Error constructing staged welcome");
+    //ANCHOR_END: not_join_group_staged_welcome
+
+    // check storage state after staging welcome
+    let own_leaf_nodes: Vec<LeafNode> = bob_provider.storage().own_leaf_nodes(&group_id).unwrap();
+    // TODO: is this correct?
+    assert!(own_leaf_nodes.is_empty());
+
+    let own_leaf_index: Option<LeafNodeIndex> =
+        bob_provider.storage().own_leaf_index(&group_id).unwrap();
+
+    // TODO: is this correct?
+    assert!(own_leaf_index.is_none());
+
+    //ANCHOR: not_join_group_welcome_sender
+    let welcome_sender: &LeafNode = staged_welcome
+        .welcome_sender()
+        .expect("Welcome sender could not be retrieved");
+
+    // Inspect sender's credential...
+    let _credential = welcome_sender.credential();
+    // Inspect sender's signature public key...
+    let _signature_key = welcome_sender.signature_key();
+    // Inspect sender's encryption public key...
+    let _encryption_key = welcome_sender.encryption_key();
```

Comment:
The encryption key is not really interesting for the application. At least I can't think of a case where it should be.